lib/request.js

```diff
@@ -5,6 +5,7 @@ function Request (params, req, body, query, log) {
   this.req = req
   this.body = body
   this.query = query
+  this.headers = (req && req.headers) ? req.headers : {}
```

Comment:
In which cases `req` could be `undefined`?

Reply:
> Correct me If I'm wrong but with your approach, I get lots of Cannot read property headers of undefined errors because at this place https://github.com/StarpTech/fastify/blob/feature/%23200_%23191_headers_shortcut_schema_validation/fastify.js#L248 we build a request object without passing the parameters. I think this construct belongs to the performance optimization. Passing headers as an argument work fine.nstruct belongs to the performance optimization. Passing headers as an argument work fine.

Reply:
@delvedor it should only be in tests where things like `'request'` is being passed as the request.

Reply:
Ok, the problem is that you are calling `Request` (the `new R()` inside `buildRequest`) without parameters, so it throws an error.
I think we should pass `req.headers` since we are also passing `req.log` for the same reason.

The `buildRequest` function is called only during the startup phase, while the `Request` constructor is always called during a request execution. So we can avoid that useless `req && req.headers`. (cc @jsumners)

Reply:
@jsumners no also when I run a single GET test the issue appears.

Reply:
As previosly mentioned the reason is here https://github.com/StarpTech/fastify/blob/feature/%23200_%23191_headers_shortcut_schema_validation/fastify.js#L248 you can even reproduce it if you just start the server

Reply:
@delvedor so I can revert my changes?

Reply:
```js
// lib/request.js 
function Request (params, req, body, query, headers, log) {
  this.params = params
  this.req = req
  this.body = body
  this.query = query
  this.headers = headers
  this.log = log
}

function buildRequest (R) {
  function _Request (params, req, body, query, headers, log) {
    this.params = params
    this.req = req
    this.body = body
    this.query = query
    this.headers = headers
    this.log = log
  }
  _Request.prototype = new R()
  return _Request
}
```

```js
// lib/handleRequest.js
var request = new store.Request(params, req, body, query, req.headers, req.log)
```

Reply:
@delvedor I revert the changes so `headers` is passed like `req.log` to avoid the checks